xpublish_wms/getmap.py

```diff
@@ -0,0 +1,345 @@
+
+        Otherwise:
+            - Get latest one
+
+        :param da:
+        :return:
+        """
+        if self.time is not None:
+            da = da.cf.sel(
+                {self.TIME_CF_NAME: self.time},
+                method="nearest"
+            )
+        else:
+            da = da.cf.isel({
+                self.TIME_CF_NAME: -1
+            })
+
+        return da
+
+    def select_custom_dim(self, da: xr.DataArray) -> xr.DataArray:
+        """
+        Select other dimension, ensuring a 2D array
+        :param da:
+        :return:
+        """
+        # Squeeze single value dimensions
+        da = da.squeeze()
+
+        # TODO: Filter dimension from custom query, if any
+
+        # Squeeze multiple values dimensions, by selecting the last value
+        for coord_name in da.cf.coords:
+            if coord_name in ("latitude", "longitude"):
+                continue
+            coord = da.cf.coords[coord_name]
+            if coord.size > 1:
+                da = da.cf.isel({coord_name: -1})
+
+        return da
+
+    def select_grid(self, da: xr.DataArray) -> xr.DataArray:
+        """
+        Select grid and reproject if needed
+        :param da:
+        :return:
+        """
+        # Some basic check for dataset
+        if not da.rio.crs:
+            da = da.rio.write_crs(self.DEFAULT_CRS)
+
+        if self.grid_is_regular(da):
+            da = self.select_regular_grid(da)
+        else:
+            da = self.select_irregular_grid(da)
+        return da
+
+    def grid_is_regular(self, da: xr.DataArray) -> bool:
+        return da.cf.coords['longitude'].dims[0] == da.cf.coords['longitude'].name
+
+    def select_regular_grid(self, da: xr.DataArray) -> xr.DataArray:
+        """
+        Filter regular grid according to WMS request :
+            - bbox
+            - width and height
+        :param da:
+        :return:
+        """
+        # Regular grid
+        minx, miny, maxx, maxy = self.bbox
+        if "4326" in self.crs:
+            # Something's strange with bbox ?
+            miny, minx, maxy, maxx = self.bbox
+
+        # TBD:
+        #   got issues on my dataset using rioxarray :
+        #       The WKT could not be parsed. OGR Error code 5
+        #   use xr.sel as backup
+
+        #clipped = da.rio.clip_box(*self.bbox, crs=self.crs)
+        #resampled_data = clipped.rio.reproject(
+        #    dst_crs=self.crs,
+        #    shape=(self.width, self.height),
+        #    resampling=Resampling.nearest,
+        #    transform=from_bounds(*self.bbox, width=self.width, height=self.height),
+        #)
+        #
+
+        longitude = np.linspace(minx, maxx, self.width)
+        latitude = np.linspace(miny, maxy, self.height)
+
+        if da.longitude[0] > da.longitude[-1]:
+            longitude = np.flip(longitude)
+        if da.latitude[0] > da.latitude[-1]:
+            latitude = np.flip(latitude)
+
+        resampled_data = da.cf.interp(
```

Comment:
One note on this, that if the output CRS is not the same as the input, the results need to be warped, which is why we are using rasterio, otherwise Latitude will not line up for Mercator projections. The time overhead of rasterio is pretty small compared to the open_dataset calls themselves